File: cses.fi/2422.cpp
#include <iostream>

using namespace std;

#define ll long long

ll count_in_matrix(ll x, ll n) {
    ll count = 0;

    for (int i = 1; i <= n; ++i) {
        count += min(x / i, n);
    }

    return count;
}

int main() {
    ll n;
    cin >> n;
    ll middle_elem = (n * n - 1) / 2;

    ll left = 1, right = n * n;

    while (left < right) {
        ll m = left + (right - left) / 2;

        if (count_in_matrix(m, n) <= middle_elem) {
            left = m  + 1;
        } else {
            right = m;
        }
    }

    cout << left << endl;
}

File: cses.fi/2422.py
def count_less_equal(x, n):
    count = 0
    for i in range(1, n + 1):
        # Count elements in the i-th row that are <= x
        count += min(x // i, n)
    return count

def find_middle_element(n):
    left, right = 1, n * n
    middle_position = (n * n - 1) // 2

    while left < right:
        mid = (left + right) // 2
        if count_less_equal(mid, n) <= middle_position:
            left = mid + 1
        else:
            right = mid

    return left

n = int(input())
print(find_middle_element(n))